provider/ec2/instancetypes.go

```diff
@@ -0,0 +1,328 @@
+			Filters:      []*ec2.Filter{zoneFilter},
+		})
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, offering := range offeringResults.InstanceTypeOfferings {
+			if _, ok := instanceTypeRegions[*offering.InstanceType]; !ok {
+				instanceTypeRegions[*offering.InstanceType] = set.NewStrings()
+			}
+			instanceTypeRegions[*offering.InstanceType].Add(*offering.Location)
+			instTypeNames = append(instTypeNames, offering.InstanceType)
+		}
+		token = offeringResults.NextToken
+		if token == nil {
+			break
+		}
+	}
+
+	// Populate the costs fo the instance types in use.
+	costs := make(map[string]uint64)
+	if err := instanceTypeCosts(ec2Session, instTypeNames, zoneNames, costs); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Compose the results.
+	var allInstanceTypes []instances.InstanceType
+	for {
+		querySize := len(instTypeNames)
+		if querySize > 100 {
+			querySize = 100
+		}
+		instTypeParams := &ec2.DescribeInstanceTypesInput{
+			InstanceTypes: instTypeNames[0:querySize],
+		}
+		instTypeResults, err := ec2Session.DescribeInstanceTypes(instTypeParams)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, info := range instTypeResults.InstanceTypes {
+			instType := instances.InstanceType{
+				Name:       *info.InstanceType,
+				VirtType:   virtType(info),
+				Deprecated: info.CurrentGeneration == nil || !*info.CurrentGeneration,
+				Cost:       costs[*info.InstanceType],
+			}
+			if info.VCpuInfo != nil && info.VCpuInfo.DefaultVCpus != nil {
+				instType.CpuCores = uint64(*info.VCpuInfo.DefaultVCpus)
+				if info.ProcessorInfo != nil && info.ProcessorInfo.SustainedClockSpeedInGhz != nil {
+					cpupower := calculateCPUPower(
+						instType.Name,
+						info.ProcessorInfo.SustainedClockSpeedInGhz,
+						uint64(*info.VCpuInfo.DefaultVCpus))
+					instType.CpuPower = &cpupower
+				}
+			}
+			if info.MemoryInfo != nil && info.MemoryInfo.SizeInMiB != nil {
+				instType.Mem = uint64(*info.MemoryInfo.SizeInMiB)
+			}
+			if info.ProcessorInfo != nil {
+				for _, instArch := range info.ProcessorInfo.SupportedArchitectures {
+					instType.Arches = append(instType.Arches, archName(*instArch))
+				}
+			}
+			if instRegions, ok := instanceTypeRegions[instType.Name]; !ok || instRegions.Size() < len(zoneNames) {
+				instType.Deprecated = true
+			}
+			allInstanceTypes = append(allInstanceTypes, instType)
+		}
+		if len(instTypeNames) <= 100 {
+			break
+		}
+		instTypeNames = instTypeNames[querySize:]
+	}
+
+	if isVPCIDSet(e.ecfg().vpcID()) {
+		return allInstanceTypes, nil
+	}
+	hasDefaultVPC, err := e.hasDefaultVPC(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if hasDefaultVPC {
+		return allInstanceTypes, nil
+	}
+
+	// The region has no default VPC, and the user has not specified
+	// one to use. We filter out any instance types that are not
+	// supported in EC2-Classic.
+	supportedInstanceTypes := make([]instances.InstanceType, 0, len(allInstanceTypes))
```

Comment:
poor allocation, I'd prefer if we had a numClassic() func or const that can seed the preallocation size here.

Reply:
yeah, i just copied this and didn't notice.
we're talking small numbers, easier just to append to an initially empty slice